src/distance/edge-filters.hpp

```diff
@@ -0,0 +1,79 @@
+#ifndef SRC_DISTANCE_EDGE_FILTERS_HPP_
+#define SRC_DISTANCE_EDGE_FILTERS_HPP_
+
+#include <memory>
+#include <optional>
+
+#include "command-line/parsing/options.hpp"
+#include "common/pipeline/pipelines.hpp"
+#include "distance/edge.hpp"
+
+namespace found {
+
+
+using EdgeFilteringAlgorithms = ModifyingPipeline<Points>;
+
+/**
+ * Abstract superclass for edge-filtering algorithms.
+ * modifies Points in-place (ModifyingStage<Points>).
+ */
+class EdgeFilteringAlgorithm : public ModifyingStage<Points> {
+ public:
+    EdgeFilteringAlgorithm() = default;
+    virtual ~EdgeFilteringAlgorithm() = default;
+    // Implementations override void Run(Points &)
+};
+
+/**
+ * A trivial implementation to ensure linkage / prevent undefined references.
+ * Does nothing to the Points (no-op filter).
+ */
+class NoOpEdgeFilter : public EdgeFilteringAlgorithm {
+ public:
+    NoOpEdgeFilter() = default;
+    ~NoOpEdgeFilter() override = default;
+
+    void Run(Points &pts) override {
+        // intentionally no-op
+        (void)pts;
+    }
+};
+
+namespace detail {
+inline NoOpEdgeFilter kNoOpEdgeFilter{};
+}  // namespace detail
+
+/**
+ * Convenience provider (no-arg). Returns a pipeline that contains a single
+ * NoOpEdgeFilter so callers that expect a ready pipeline can use this.
+ */
+inline EdgeFilteringAlgorithms ProvideEdgeFilteringAlgorithm() {
+    EdgeFilteringAlgorithms pipeline;
+    pipeline.Complete(detail::kNoOpEdgeFilter);
+    return pipeline;
+}
+
+/**
+ * Options-aware provider. Constructs a pipeline containing all filters
+ * that are enabled in options and returns it. If no filter is enabled,
+ * returns std::nullopt to indicate "no filters".
+ *
+ * Note: stages added are function-local static instances so the pipeline
+ * can safely store raw pointers to them.
+ */
+inline std::optional<EdgeFilteringAlgorithms> ProvideEdgeFilteringAlgorithm(const DistanceOptions &options) {
```

Comment:
Return type should be unique_ptr, not `std::optional`. If you think about it, it should never be optional, because our executor constructor doesn't accept an optional object. That's why we overload the constructor, because we do know at runtime which one to call and whether it the EdgeFilteringAlgorithms will exist.